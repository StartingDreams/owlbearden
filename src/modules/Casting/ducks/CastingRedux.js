import { fromJS } from 'immutable';
import firebase from 'firebase';

const database = firebase.database();

export const Types = {
  GET_REQUEST: 'casting/GET_REQUEST',
  GET_SUCCESS: 'casting/GET_SUCCESS',
  GET_FAILURE: 'casting/GET_FAILURE',
  SET_FILTERS: 'casting/SET_FILTERS',
  SET_FILTERED_SPELLS: 'casting/SET_FILTERED_SPELLS',
};

export const getRequest = () => ({
  type: Types.GET_REQUEST,
});

export const getSuccess = allSpells => ({
  type: Types.GET_SUCCESS,
  allSpells,
});

export const getFailure = error => ({
  type: Types.GET_FAILURE,
  error,
});

export const setFilteredSpells = filteredSpells => ({
  type: Types.SET_FILTERED_SPELLS,
  filteredSpells,
});

export const setFilters = filters => ({
  type: Types.SET_FILTERS,
  filters,
});

export const getAllSpells = (dispatch) => {
  try {
    dispatch(getRequest());
    database.ref('dnd/casting/spells/').once('value', (snapshot) => {
      const values = snapshot.val();
      const rawSpells = Object.keys(values).map(key => (values[key]));
      const allSpells = fromJS(rawSpells);
      dispatch(getSuccess(allSpells));
    }, (error) => {
      dispatch(getFailure(error.message));
    });
  } catch (error) {
    dispatch(getFailure(error.message));
  }
};

export const INITIAL_STATE = fromJS({
  fetching: false,
  redirecting: false,
  filteredSpells: [],
  allSpells: [],
  filters: {
    levels: [1],
    classes: ['Wizard'],
    range: [],
    concentration: [],
    components: [],
    ritual: [],
  },
});

export default function reducer(state = INITIAL_STATE, action = {}) {
  switch (action.type) {
    case Types.GET_REQUEST:
      return state.set('fetching', true);
    case Types.GET_SUCCESS:
      return state.withMutations(s => s
        .set('allSpells', fromJS(action.allSpells)))
        .set('fetching', false);
    case Types.GET_FAILURE:
      return state.set('fetching', false);
    case Types.SET_FILTERED_SPELLS:
      return state.withMutations(s => s
        .set('filteredSpells', fromJS(action.filteredSpells)));
    case Types.SET_FILTERS:
      return state.set('filters', action.filters);
    default:
      return state;
  }
}
