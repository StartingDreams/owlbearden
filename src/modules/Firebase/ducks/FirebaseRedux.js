import { fromJS } from 'immutable';

export const INITIAL_STATE = fromJS({});

export default function reducer(state = INITIAL_STATE, action = {}) {
  switch (action.type) {
    default:
      return state;
  }
}
