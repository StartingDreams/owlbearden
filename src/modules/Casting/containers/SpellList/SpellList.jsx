import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ImmutablePropTypes from 'react-immutable-proptypes';
import * as CastingActions from '../../ducks/CastingRedux';
import Spell from '../../components/Spell';
import FilterBar from '../FilterBar';

export class SpellList extends React.Component {
  componentDidMount() {
    const { dispatch, allSpells } = this.props;
    if (allSpells.size !== 0) {
      return;
    }
    CastingActions.getAllSpells(dispatch);
  }

  shouldComponentUpdate(nextProps) {
    if (!nextProps.filteredSpells || nextProps.filteredSpells === 'undefined') {
      return false;
    }
    if (this.props.allSpells !== nextProps.allSpells) {
      return true;
    }
    return this.props.filteredSpells !== nextProps.filteredSpells;
  }

  render() {
    const { filteredSpells, allSpells } = this.props;
    return (
      <div>
        { allSpells.size === 0 ? null : <FilterBar /> }
        { filteredSpells.map((spell, key) => (
          <Spell
            key={key}
            spell={spell}
          />
        ))}
      </div>
    );
  }
}

SpellList.propTypes = {
  dispatch: PropTypes.func.isRequired,
  allSpells: ImmutablePropTypes.listOf(
    ImmutablePropTypes.contains({
      name: PropTypes.string.isRequired,
    }),
  ).isRequired,
  filteredSpells: ImmutablePropTypes.listOf(
    ImmutablePropTypes.contains({
      name: PropTypes.string.isRequired,
    }),
  ),
};

export const mapStateToProps = state => ({
  filteredSpells: state.casting.get('filteredSpells'),
  allSpells: state.casting.get('allSpells'),
});

export default connect(
  mapStateToProps,
)(SpellList);
