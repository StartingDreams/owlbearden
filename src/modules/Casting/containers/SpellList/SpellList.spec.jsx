import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { fromJS } from 'immutable';
import renderer from 'react-test-renderer';
import { SpellList, mapStateToProps } from './SpellList';
import * as CastingActions from '../../ducks/CastingRedux';

jest.mock('../../ducks/CastingRedux', () => ({ getAllSpells: jest.fn() }));
jest.mock('../../components/Spell', () => (() => (null)));
jest.mock('../FilterBar', () => (() => (null)));

const spells = [
  {
    casting_time: '1 action',
    classes: [{
      name: 'Wizard',
      url: 'http://www.dnd5eapi.co/api/classes/12',
    }],
    components: ['V', 'S', 'M'],
    concentration: 'no',
    desc: ['A shimmering green arrow streaks toward a target within range and bursts in a spray of acid. Make a ranged spell attack against the target. On a hit, the target takes 4d4 acid damage immediately and 2d4 acid damage at the end of its next turn. On a miss, the arrow splashes the target with acid for half as much of the initial damage and no damage at the end of its next turn.'],
    duration: 'Instantaneous',
    higher_level: ['When you cast this spell using a spell slot of 3rd level or higher, the damage (both initial and later) increases by 1d4 for each slot level above 2nd.'],
    id: '58f40b80c9e7ce9f72153112',
    index: 1,
    level: 2,
    material: 'Powdered rhubarb leaf and an adder’s stomach.',
    name: 'Acid Arrow',
    page: 'phb 259',
    range: '90 feet',
    ritual: 'no',
    school: {
      name: 'Evocation',
      url: 'http://www.dnd5eapi.co/api/magic-schools/5',
    },
    subclasses: [{
      name: 'Lore',
      url: 'http://www.dnd5eapi.co/api/subclasses/2',
    }, {
      name: 'Land',
      url: 'http://www.dnd5eapi.co/api/subclasses/4',
    }],
    url: 'http://www.dnd5eapi.co/api/spells/1',
  },
  {
    casting_time: '1 minute',
    classes: [{
      name: 'Cleric',
      url: 'http://www.dnd5eapi.co/api/classes/3',
    }],
    components: ['V', 'S', 'M'],
    concentration: 'no',
    desc: ['You contact your deity or a divine proxy and ask up to three questions that can be answered with a yes or no. You must ask your questions before the spell ends. You receive a correct answer for each question.', 'Divine beings aren’t necessarily omniscient, so you might receive “unclear” as an answer if a question pertains to information that lies beyond the deity’s knowledge. In a case where a one-word answer could be misleading or contrary to the deity’s interests, the DM might offer a short phrase as an answer instead.', 'If you cast the spell two or more times before finishing your next long rest, there is a cumulative 25 percent chance for each casting after the first that you get no answer. The DM makes this roll in secret.'],
    duration: '1 minute',
    id: '58f40b80c9e7ce9f72153149',
    index: 45,
    level: 5,
    material: 'Incense and a vial of holy or unholy water.',
    name: 'Commune',
    page: 'phb 223',
    range: 'Self',
    ritual: 'yes',
    school: {
      name: 'Divination',
      url: 'http://www.dnd5eapi.co/api/magic-schools/3',
    },
    subclasses: [{
      name: 'Devotion',
      url: 'http://www.dnd5eapi.co/api/subclasses/7',
    }],
    url: 'http://www.dnd5eapi.co/api/spells/45',
  },
  {
    casting_time: '1 bonus action',
    classes: [{
      name: 'Paladin',
      url: 'http://www.dnd5eapi.co/api/classes/7',
    }],
    components: ['V', 'S'],
    concentration: 'yes',
    desc: ['Your prayer empowers you with divine radiance. Until the spell ends, your weapon attacks deal an extra 1d4 radiant damage on a hit.'],
    duration: 'Up to 1 minute',
    id: '58f40b80c9e7ce9f72153164',
    index: 84,
    level: 1,
    name: 'Divine Favor',
    page: 'phb 234',
    range: 'Self',
    ritual: 'no',
    school: {
      name: 'Evocation',
      url: 'http://www.dnd5eapi.co/api/magic-schools/5',
    },
    subclasses: [{
      name: 'Lore',
      url: 'http://www.dnd5eapi.co/api/subclasses/2',
    }],
    url: 'http://www.dnd5eapi.co/api/spells/84',
  },
];

const immutableSpells = fromJS(spells);
const noSpells = fromJS([]);
const testSpell = fromJS([spells[0]]);

describe('<SpellList />', () => {
  it('one spell should match the snapshot, and getAllSpells to not be called', () => {
    const dispatch = () => ('dispatch');
    renderer.create(<SpellList
      allSpells={testSpell}
      filteredSpells={testSpell}
      dispatch={dispatch}
    />);
    expect(CastingActions.getAllSpells.mock.calls.length).toBe(0);
  });
  it('zero spells should match the snapshot and fetch more spells', () => {
    const dispatch = () => ('dispatch');
    renderer.create(<SpellList
      allSpells={noSpells}
      filteredSpells={noSpells}
      dispatch={dispatch}
    />);
    expect(CastingActions.getAllSpells).toBeCalledWith(dispatch);
  });
  it('two spells should match the snapshot', () => {
    const dispatch = () => ('dispatch');
    const wrapper = shallow(<SpellList
      allSpells={immutableSpells}
      filteredSpells={immutableSpells}
      dispatch={dispatch}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('mapStateToProps - spells should be passed in', () => {
    const casting = fromJS({
      spells,
    });
    const state = {
      casting,
    };
    const props = mapStateToProps(state);
    expect(props.allSpells).toBe(casting.get('allSpells'));
  });
});
