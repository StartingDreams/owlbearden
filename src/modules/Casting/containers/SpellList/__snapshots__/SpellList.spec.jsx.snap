// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpellList /> two spells should match the snapshot 1`] = `
<div>
  <Component />
  <Component
    spell={
      Immutable.Map {
        ritual: "no",
        level: 2,
        classes: Immutable.List [
          Immutable.Map {
            name: "Wizard",
            url: "http://www.dnd5eapi.co/api/classes/12",
          },
        ],
        subclasses: Immutable.List [
          Immutable.Map {
            name: "Lore",
            url: "http://www.dnd5eapi.co/api/subclasses/2",
          },
          Immutable.Map {
            name: "Land",
            url: "http://www.dnd5eapi.co/api/subclasses/4",
          },
        ],
        material: "Powdered rhubarb leaf and an adder’s stomach.",
        casting_time: "1 action",
        name: "Acid Arrow",
        concentration: "no",
        url: "http://www.dnd5eapi.co/api/spells/1",
        page: "phb 259",
        desc: Immutable.List [
          "A shimmering green arrow streaks toward a target within range and bursts in a spray of acid. Make a ranged spell attack against the target. On a hit, the target takes 4d4 acid damage immediately and 2d4 acid damage at the end of its next turn. On a miss, the arrow splashes the target with acid for half as much of the initial damage and no damage at the end of its next turn.",
        ],
        index: 1,
        higher_level: Immutable.List [
          "When you cast this spell using a spell slot of 3rd level or higher, the damage (both initial and later) increases by 1d4 for each slot level above 2nd.",
        ],
        duration: "Instantaneous",
        school: Immutable.Map {
          name: "Evocation",
          url: "http://www.dnd5eapi.co/api/magic-schools/5",
        },
        components: Immutable.List [
          "V",
          "S",
          "M",
        ],
        id: "58f40b80c9e7ce9f72153112",
        range: "90 feet",
      }
    }
  />
  <Component
    spell={
      Immutable.Map {
        ritual: "yes",
        level: 5,
        classes: Immutable.List [
          Immutable.Map {
            name: "Cleric",
            url: "http://www.dnd5eapi.co/api/classes/3",
          },
        ],
        subclasses: Immutable.List [
          Immutable.Map {
            name: "Devotion",
            url: "http://www.dnd5eapi.co/api/subclasses/7",
          },
        ],
        material: "Incense and a vial of holy or unholy water.",
        casting_time: "1 minute",
        name: "Commune",
        concentration: "no",
        url: "http://www.dnd5eapi.co/api/spells/45",
        page: "phb 223",
        desc: Immutable.List [
          "You contact your deity or a divine proxy and ask up to three questions that can be answered with a yes or no. You must ask your questions before the spell ends. You receive a correct answer for each question.",
          "Divine beings aren’t necessarily omniscient, so you might receive “unclear” as an answer if a question pertains to information that lies beyond the deity’s knowledge. In a case where a one-word answer could be misleading or contrary to the deity’s interests, the DM might offer a short phrase as an answer instead.",
          "If you cast the spell two or more times before finishing your next long rest, there is a cumulative 25 percent chance for each casting after the first that you get no answer. The DM makes this roll in secret.",
        ],
        index: 45,
        duration: "1 minute",
        school: Immutable.Map {
          name: "Divination",
          url: "http://www.dnd5eapi.co/api/magic-schools/3",
        },
        components: Immutable.List [
          "V",
          "S",
          "M",
        ],
        id: "58f40b80c9e7ce9f72153149",
        range: "Self",
      }
    }
  />
  <Component
    spell={
      Immutable.Map {
        ritual: "no",
        level: 1,
        classes: Immutable.List [
          Immutable.Map {
            name: "Paladin",
            url: "http://www.dnd5eapi.co/api/classes/7",
          },
        ],
        subclasses: Immutable.List [
          Immutable.Map {
            name: "Lore",
            url: "http://www.dnd5eapi.co/api/subclasses/2",
          },
        ],
        casting_time: "1 bonus action",
        name: "Divine Favor",
        concentration: "yes",
        url: "http://www.dnd5eapi.co/api/spells/84",
        page: "phb 234",
        desc: Immutable.List [
          "Your prayer empowers you with divine radiance. Until the spell ends, your weapon attacks deal an extra 1d4 radiant damage on a hit.",
        ],
        index: 84,
        duration: "Up to 1 minute",
        school: Immutable.Map {
          name: "Evocation",
          url: "http://www.dnd5eapi.co/api/magic-schools/5",
        },
        components: Immutable.List [
          "V",
          "S",
        ],
        id: "58f40b80c9e7ce9f72153164",
        range: "Self",
      }
    }
  />
</div>
`;
